// find-and-replace/lib/history.js
(function() {
  var HISTORY_MAX, History, _;

  _ = require('underscore-plus');

  HISTORY_MAX = 25;

  module.exports = History = (function() {
    function History(miniEditor, items) {
      this.miniEditor = miniEditor;
      this.items = items != null ? items : [];
      this.index = this.items.length;
      this.miniEditor.on('core:move-up', (function(_this) {
        return function() {
          return _this.previous();
        };
      })(this));
      this.miniEditor.on('core:move-down', (function(_this) {
        return function() {
          return _this.next();
        };
      })(this));
    }

    History.prototype.serialize = function() {
      return this.items.slice(-HISTORY_MAX);
    };

    History.prototype.previous = function() {
      var _ref;
      if (this.items.length === 0 || (this.atLastItem() && this.miniEditor.getText() !== this.getLast())) {
        this.scratch = this.miniEditor.getText();
      } else if (this.index > 0) {
        this.index--;
      }
      return this.miniEditor.setText((_ref = this.items[this.index]) != null ? _ref : '');
    };

    History.prototype.next = function() {
      var item;
      if (this.index < this.items.length - 1) {
        this.index++;
        item = this.items[this.index];
      } else if (this.scratch) {
        item = this.scratch;
      } else {
        item = '';
      }
      return this.miniEditor.setText(item);
    };

    History.prototype.getLast = function() {
      return _.last(this.items);
    };

    History.prototype.atLastItem = function() {
      return this.index === this.items.length - 1;
    };

    History.prototype.store = function() {
      var text;
      text = this.miniEditor.getText();
      if (!text || text === this.getLast()) {
        return;
      }
      this.scratch = null;
      this.items.push(text);
      return this.index = this.items.length - 1;
    };

    return History;

  })();

}).call(this);
