// fuzzy-finder/lib/buffer-view.js
(function() {
  var BufferView, FuzzyFinderView, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  _ = require('underscore-plus');

  FuzzyFinderView = require('./fuzzy-finder-view');

  module.exports = BufferView = (function(_super) {
    __extends(BufferView, _super);

    function BufferView() {
      return BufferView.__super__.constructor.apply(this, arguments);
    }

    BufferView.prototype.toggle = function() {
      var _ref;
      if (this.hasParent()) {
        return this.cancel();
      } else {
        this.allowActiveEditorChange = true;
        this.populate();
        if (((_ref = this.paths) != null ? _ref.length : void 0) > 0) {
          return this.attach();
        }
      }
    };

    BufferView.prototype.getEmptyMessage = function(itemCount) {
      if (itemCount === 0) {
        return 'No open editors';
      } else {
        return BufferView.__super__.getEmptyMessage.apply(this, arguments);
      }
    };

    BufferView.prototype.populate = function() {
      var editor, editors, _i, _len;
      editors = atom.project.getEditors().filter(function(editor) {
        return editor.getPath() != null;
      });
      editors = _.sortBy(editors, function(editor) {
        if (editor === atom.workspaceView.getActivePaneItem()) {
          return 0;
        } else {
          return -(editor.lastOpened || 1);
        }
      });
      this.paths = [];
      for (_i = 0, _len = editors.length; _i < _len; _i++) {
        editor = editors[_i];
        this.paths.push(editor.getPath());
      }
      return this.setItems(_.uniq(this.paths));
    };

    return BufferView;

  })(FuzzyFinderView);

}).call(this);
