// fuzzy-finder/lib/fuzzy-finder.js
(function() {
  module.exports = {
    configDefaults: {
      ignoredNames: []
    },
    activate: function(state) {
      var PathLoader, editor, _i, _len, _ref;
      atom.workspaceView.command('fuzzy-finder:toggle-file-finder', (function(_this) {
        return function() {
          return _this.createProjectView().toggle();
        };
      })(this));
      atom.workspaceView.command('fuzzy-finder:toggle-buffer-finder', (function(_this) {
        return function() {
          return _this.createBufferView().toggle();
        };
      })(this));
      atom.workspaceView.command('fuzzy-finder:toggle-git-status-finder', (function(_this) {
        return function() {
          return _this.createGitStatusView().toggle();
        };
      })(this));
      if (atom.project.getPath() != null) {
        PathLoader = require('./path-loader');
        this.loadPathsTask = PathLoader.startTask((function(_this) {
          return function(paths) {
            return _this.projectPaths = paths;
          };
        })(this));
      }
      _ref = atom.project.getEditors();
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        editor = _ref[_i];
        editor.lastOpened = state[editor.getPath()];
      }
      return atom.workspaceView.eachPane(function(pane) {
        var _ref1;
        if ((_ref1 = pane.activeItem) != null) {
          _ref1.lastOpened = Date.now();
        }
        return pane.on('pane:active-item-changed', function(e, item) {
          return item.lastOpened = Date.now();
        });
      });
    },
    deactivate: function() {
      if (this.loadPathsTask != null) {
        this.loadPathsTask.terminate();
        this.loadPathsTask = null;
      }
      if (this.projectView != null) {
        this.projectView.destroy();
        this.projectView = null;
      }
      if (this.bufferView != null) {
        this.bufferView.destroy();
        this.bufferView = null;
      }
      if (this.gitStatusView != null) {
        this.gitStatusView.destroy();
        this.gitStatusView = null;
      }
      return this.projectPaths = null;
    },
    serialize: function() {
      var editor, path, paths, _i, _len, _ref;
      paths = {};
      _ref = atom.project.getEditors();
      for (_i = 0, _len = _ref.length; _i < _len; _i++) {
        editor = _ref[_i];
        path = editor.getPath();
        if (path != null) {
          paths[path] = editor.lastOpened;
        }
      }
      return paths;
    },
    createProjectView: function() {
      var ProjectView, _ref;
      if (this.projectView == null) {
        if ((_ref = this.loadPathsTask) != null) {
          _ref.terminate();
        }
        ProjectView = require('./project-view');
        this.projectView = new ProjectView(this.projectPaths);
        this.projectPaths = null;
      }
      return this.projectView;
    },
    createGitStatusView: function() {
      var GitStatusView;
      if (this.gitStatusView == null) {
        GitStatusView = require('./git-status-view');
        this.gitStatusView = new GitStatusView();
      }
      return this.gitStatusView;
    },
    createBufferView: function() {
      var BufferView;
      if (this.bufferView == null) {
        BufferView = require('./buffer-view');
        this.bufferView = new BufferView();
      }
      return this.bufferView;
    }
  };

}).call(this);
