// fuzzy-finder/lib/project-view.js
(function() {
  var $, FuzzyFinderView, PathLoader, ProjectView, humanize,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  $ = require('atom').$;

  humanize = require('humanize-plus');

  FuzzyFinderView = require('./fuzzy-finder-view');

  PathLoader = require('./path-loader');

  module.exports = ProjectView = (function(_super) {
    __extends(ProjectView, _super);

    function ProjectView() {
      return ProjectView.__super__.constructor.apply(this, arguments);
    }

    ProjectView.prototype.paths = null;

    ProjectView.prototype.reloadPaths = true;

    ProjectView.prototype.reloadAfterFirstLoad = false;

    ProjectView.prototype.initialize = function(paths) {
      var _ref;
      this.paths = paths;
      ProjectView.__super__.initialize.apply(this, arguments);
      if (((_ref = this.paths) != null ? _ref.length : void 0) > 0) {
        this.reloadPaths = false;
      }
      this.subscribe($(window), 'focus', (function(_this) {
        return function() {
          if (_this.paths != null) {
            return _this.reloadPaths = true;
          } else {
            return _this.reloadAfterFirstLoad = true;
          }
        };
      })(this));
      return this.subscribe(atom.config.observe('fuzzy-finder.ignoredNames', {
        callNow: false
      }, (function(_this) {
        return function() {
          return _this.reloadPaths = true;
        };
      })(this)));
    };

    ProjectView.prototype.toggle = function() {
      if (this.hasParent()) {
        return this.cancel();
      } else if (atom.project.getPath() != null) {
        this.populate();
        return this.attach();
      }
    };

    ProjectView.prototype.getEmptyMessage = function(itemCount) {
      if (itemCount === 0) {
        return 'Project is empty';
      } else {
        return ProjectView.__super__.getEmptyMessage.apply(this, arguments);
      }
    };

    ProjectView.prototype.populate = function() {
      var pathsFound, _ref;
      if (this.paths != null) {
        this.setItems(this.paths);
      }
      if (this.reloadPaths) {
        this.reloadPaths = false;
        if ((_ref = this.loadPathsTask) != null) {
          _ref.terminate();
        }
        this.loadPathsTask = PathLoader.startTask((function(_this) {
          return function(paths) {
            _this.paths = paths;
            if (_this.reloadAfterFirstLoad) {
              _this.reloadPaths = true;
              _this.reloadAfterFirstLoad = false;
            }
            return _this.populate();
          };
        })(this));
        if (this.paths != null) {
          return this.setLoading("Reindexing project\u2026");
        } else {
          this.setLoading("Indexing project\u2026");
          this.loadingBadge.text('0');
          pathsFound = 0;
          return this.loadPathsTask.on('load-paths:paths-found', (function(_this) {
            return function(paths) {
              pathsFound += paths.length;
              return _this.loadingBadge.text(humanize.intComma(pathsFound));
            };
          })(this));
        }
      }
    };

    ProjectView.prototype.projectRelativePathsForFilePaths = function() {
      var filePath, index, lastOpenedEntry, lastOpenedPath, projectRelativePath, projectRelativePaths, _i, _len;
      projectRelativePaths = ProjectView.__super__.projectRelativePathsForFilePaths.apply(this, arguments);
      if (lastOpenedPath = this.getLastOpenedPath()) {
        for (index = _i = 0, _len = projectRelativePaths.length; _i < _len; index = ++_i) {
          filePath = projectRelativePaths[index].filePath;
          if (filePath === lastOpenedPath) {
            projectRelativePaths.splice(index, 1);
            break;
          }
        }
        projectRelativePath = atom.project.relativize(lastOpenedPath);
        lastOpenedEntry = {
          filePath: lastOpenedPath,
          projectRelativePath: projectRelativePath
        };
        projectRelativePaths.unshift(lastOpenedEntry);
      }
      return projectRelativePaths;
    };

    ProjectView.prototype.getLastOpenedPath = function() {
      var activePath, editor, filePath, lastOpenedEditor, _i, _len, _ref, _ref1;
      activePath = (_ref = atom.workspace.activePaneItem) != null ? typeof _ref.getPath === "function" ? _ref.getPath() : void 0 : void 0;
      lastOpenedEditor = null;
      _ref1 = atom.project.getEditors();
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        editor = _ref1[_i];
        filePath = editor.getPath();
        if (!filePath) {
          continue;
        }
        if (activePath === filePath) {
          continue;
        }
        if (lastOpenedEditor == null) {
          lastOpenedEditor = editor;
        }
        if (editor.lastOpened > lastOpenedEditor.lastOpened) {
          lastOpenedEditor = editor;
        }
      }
      return lastOpenedEditor != null ? lastOpenedEditor.getPath() : void 0;
    };

    ProjectView.prototype.beforeRemove = function() {
      var _ref;
      return (_ref = this.loadPathsTask) != null ? _ref.terminate() : void 0;
    };

    return ProjectView;

  })(FuzzyFinderView);

}).call(this);
