// symbols-view/lib/main.js
(function() {
  module.exports = {
    configDefaults: {
      useEditorGrammarAsCtagsLanguage: true
    },
    activate: function() {
      atom.workspaceView.command('symbols-view:toggle-file-symbols', (function(_this) {
        return function() {
          return _this.createFileView().toggle();
        };
      })(this));
      atom.workspaceView.command('symbols-view:toggle-project-symbols', (function(_this) {
        return function() {
          return _this.createProjectView().toggle();
        };
      })(this));
      return atom.workspaceView.command('symbols-view:go-to-declaration', (function(_this) {
        return function() {
          return _this.createGoToView().toggle();
        };
      })(this));
    },
    deactivate: function() {
      if (this.fileView != null) {
        this.fileView.destroy();
        this.fileView = null;
      }
      if (this.projectView != null) {
        this.projectView.destroy();
        this.projectView = null;
      }
      if (this.goToView != null) {
        this.goToView.destroy();
        return this.goToView = null;
      }
    },
    createFileView: function() {
      var FileView;
      if (this.fileView == null) {
        FileView = require('./file-view');
        this.fileView = new FileView();
      }
      return this.fileView;
    },
    createProjectView: function() {
      var ProjectView;
      if (this.projectView == null) {
        ProjectView = require('./project-view');
        this.projectView = new ProjectView();
      }
      return this.projectView;
    },
    createGoToView: function() {
      var GoToView;
      if (this.goToView == null) {
        GoToView = require('./go-to-view');
        this.goToView = new GoToView();
      }
      return this.goToView;
    }
  };

}).call(this);
