// symbols-view/lib/symbols-view.js
(function() {
  var $$, Point, SelectListView, SymbolsView, fs, path, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  path = require('path');

  _ref = require('atom'), $$ = _ref.$$, Point = _ref.Point, SelectListView = _ref.SelectListView;

  fs = require('fs-plus');

  module.exports = SymbolsView = (function(_super) {
    __extends(SymbolsView, _super);

    function SymbolsView() {
      return SymbolsView.__super__.constructor.apply(this, arguments);
    }

    SymbolsView.activate = function() {
      return new SymbolsView;
    };

    SymbolsView.prototype.initialize = function() {
      SymbolsView.__super__.initialize.apply(this, arguments);
      return this.addClass('symbols-view overlay from-top');
    };

    SymbolsView.prototype.destroy = function() {
      this.cancel();
      return this.remove();
    };

    SymbolsView.prototype.getFilterKey = function() {
      return 'name';
    };

    SymbolsView.prototype.viewForItem = function(_arg) {
      var file, name, position;
      position = _arg.position, name = _arg.name, file = _arg.file;
      return $$(function() {
        return this.li({
          "class": 'two-lines'
        }, (function(_this) {
          return function() {
            var text;
            _this.div(name, {
              "class": 'primary-line'
            });
            if (position) {
              text = "Line " + (position.row + 1);
            } else {
              text = path.basename(file);
            }
            return _this.div(text, {
              "class": 'secondary-line'
            });
          };
        })(this));
      });
    };

    SymbolsView.prototype.getEmptyMessage = function(itemCount) {
      if (itemCount === 0) {
        return 'No symbols found';
      } else {
        return SymbolsView.__super__.getEmptyMessage.apply(this, arguments);
      }
    };

    SymbolsView.prototype.confirmed = function(tag) {
      if (tag.file && !fs.isFileSync(atom.project.resolve(tag.file))) {
        this.setError('Selected file does not exist');
        return setTimeout(((function(_this) {
          return function() {
            return _this.setError();
          };
        })(this)), 2000);
      } else {
        this.cancel();
        return this.openTag(tag);
      }
    };

    SymbolsView.prototype.openTag = function(tag) {
      var position;
      position = tag.position;
      if (!position) {
        position = this.getTagLine(tag);
      }
      if (tag.file) {
        return atom.workspaceView.open(tag.file).done((function(_this) {
          return function() {
            if (position) {
              return _this.moveToPosition(position);
            }
          };
        })(this));
      } else if (position) {
        return this.moveToPosition(position);
      }
    };

    SymbolsView.prototype.moveToPosition = function(position) {
      var editor, editorView;
      editorView = atom.workspaceView.getActiveView();
      if (editor = typeof editorView.getEditor === "function" ? editorView.getEditor() : void 0) {
        editorView.scrollToBufferPosition(position, {
          center: true
        });
        editor.setCursorBufferPosition(position);
        return editor.moveCursorToFirstCharacterOfLine();
      }
    };

    SymbolsView.prototype.attach = function() {
      this.storeFocusedElement();
      atom.workspaceView.appendToTop(this);
      return this.focusFilterEditor();
    };

    SymbolsView.prototype.getTagLine = function(tag) {
      var file, index, line, pattern, _i, _len, _ref1, _ref2;
      pattern = (_ref1 = tag.pattern) != null ? _ref1.replace(/(^^\/\^)|(\$\/$)/g, '').trim() : void 0;
      if (!pattern) {
        return;
      }
      file = atom.project.resolve(tag.file);
      if (!fs.isFileSync(file)) {
        return;
      }
      _ref2 = fs.readFileSync(file, 'utf8').split('\n');
      for (index = _i = 0, _len = _ref2.length; _i < _len; index = ++_i) {
        line = _ref2[index];
        if (pattern === line.trim()) {
          return new Point(index, 0);
        }
      }
    };

    return SymbolsView;

  })(SelectListView);

}).call(this);
