// release-notes/lib/release-notes-view.js
(function() {
  var ReleaseNotesView, View,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  View = require('atom').View;

  module.exports = ReleaseNotesView = (function(_super) {
    __extends(ReleaseNotesView, _super);

    function ReleaseNotesView() {
      return ReleaseNotesView.__super__.constructor.apply(this, arguments);
    }

    ReleaseNotesView.content = function() {
      return this.div({
        "class": 'release-notes padded pane-item',
        tabindex: -1
      }, (function(_this) {
        return function() {
          return _this.section({
            "class": 'description',
            outlet: 'description'
          });
        };
      })(this));
    };

    ReleaseNotesView.prototype.getTitle = function() {
      return 'Release Notes';
    };

    ReleaseNotesView.prototype.getUri = function() {
      return this.uri;
    };

    ReleaseNotesView.prototype.deserialize = function(_arg) {
      var releaseNotes, uri, version;
      uri = _arg.uri, version = _arg.version, releaseNotes = _arg.releaseNotes;
      return new ReleaseNotesView(uri, version, releaseNotes);
    };

    ReleaseNotesView.prototype.serialize = function() {
      return {
        deserializer: this.constructor.name,
        uri: this.uri,
        releaseNotes: this.releaseNotes,
        version: this.version
      };
    };

    ReleaseNotesView.prototype.initialize = function(uri, version, releaseNotes) {
      this.uri = uri;
      this.version = version;
      this.releaseNotes = releaseNotes;
      this.description.html(releaseNotes);
      this.description.prepend("<h1 class='section-heading'>" + this.version + "</h1>");
      if (this.version !== atom.getVersion()) {
        return this.description.append('<br><p><span class="inline-block highlight-info">To update:</span> close Atom and reopen it.</p>');
      }
    };

    return ReleaseNotesView;

  })(View);

}).call(this);
