// update-package-dependencies/lib/update-package-dependencies.js
(function() {
  var $$, BufferedProcess, _ref;

  _ref = require('atom'), $$ = _ref.$$, BufferedProcess = _ref.BufferedProcess;

  module.exports = {
    activate: function() {
      return atom.workspaceView.command('update-package-dependencies:update', (function(_this) {
        return function() {
          return _this.update();
        };
      })(this));
    },
    update: function() {
      var args, command, exit, options, view;
      view = this.createProgressView();
      atom.workspaceView.append(view);
      command = atom.packages.getApmPath();
      args = ['install'];
      options = {
        cwd: atom.project.getPath()
      };
      exit = function(code, signal) {
        var success;
        atom.workspaceView.one('core:cancel', function() {
          return view.remove();
        });
        view.empty().focus().on('focusout', function() {
          return view.remove();
        });
        success = code === 0;
        if (success) {
          return view.append($$(function() {
            return this.div({
              "class": 'text-success'
            }, 'Package dependencies updated.');
          }));
        } else {
          return view.append($$(function() {
            return this.div({
              "class": 'text-error'
            }, 'Failed to update package depencencies.');
          }));
        }
      };
      return new BufferedProcess({
        command: command,
        args: args,
        exit: exit,
        options: options
      });
    },
    createProgressView: function() {
      return $$(function() {
        return this.div({
          tabindex: -1,
          "class": 'overlay from-top'
        }, (function(_this) {
          return function() {
            _this.span({
              "class": 'loading loading-spinner-small inline-block'
            });
            return _this.span("Updating package dependencies\u2026");
          };
        })(this));
      });
    }
  };

}).call(this);
