// git-diff/lib/git-diff-view.js
(function() {
  var GitDiffView, Subscriber,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  Subscriber = require('emissary').Subscriber;

  module.exports = GitDiffView = (function() {
    Subscriber.includeInto(GitDiffView);

    GitDiffView.prototype.classes = ['git-line-added', 'git-line-modified', 'git-line-removed'];

    function GitDiffView(editorView) {
      var _ref;
      this.editorView = editorView;
      this.renderDiffs = __bind(this.renderDiffs, this);
      this.removeDiffs = __bind(this.removeDiffs, this);
      this.updateDiffs = __bind(this.updateDiffs, this);
      this.subscribeToBuffer = __bind(this.subscribeToBuffer, this);
      _ref = this.editorView, this.editor = _ref.editor, this.gutter = _ref.gutter;
      this.diffs = {};
      this.subscribe(this.editorView, 'editor:path-changed', this.subscribeToBuffer);
      this.subscribe(this.editorView, 'editor:display-updated', this.renderDiffs);
      this.subscribe(atom.project.getRepo(), 'statuses-changed', (function(_this) {
        return function() {
          _this.diffs = {};
          return _this.scheduleUpdate();
        };
      })(this));
      this.subscribe(atom.project.getRepo(), 'status-changed', (function(_this) {
        return function(path) {
          delete _this.diffs[path];
          if (path === _this.editor.getPath()) {
            return _this.scheduleUpdate();
          }
        };
      })(this));
      this.subscribeToBuffer();
      this.subscribe(this.editorView, 'editor:will-be-removed', (function(_this) {
        return function() {
          _this.unsubscribe();
          return _this.unsubscribeFromBuffer();
        };
      })(this));
      this.subscribeToCommand(this.editorView, 'git-diff:move-to-next-diff', (function(_this) {
        return function() {
          return _this.moveToNextDiff();
        };
      })(this));
      this.subscribeToCommand(this.editorView, 'git-diff:move-to-previous-diff', (function(_this) {
        return function() {
          return _this.moveToPreviousDiff();
        };
      })(this));
      this.subscribe(atom.config.observe('git-diff.showIconsInEditorGutter', (function(_this) {
        return function() {
          if (atom.config.get('git-diff.showIconsInEditorGutter')) {
            return _this.gutter.addClass('git-diff-icon');
          } else {
            return _this.gutter.removeClass('git-diff-icon');
          }
        };
      })(this)));
    }

    GitDiffView.prototype.moveToNextDiff = function() {
      var cursorLineNumber, hunks, newStart, nextDiffLineNumber, _i, _len, _ref;
      cursorLineNumber = this.editor.getCursorBufferPosition().row + 1;
      nextDiffLineNumber = null;
      hunks = (_ref = this.diffs[this.editor.getPath()]) != null ? _ref : [];
      for (_i = 0, _len = hunks.length; _i < _len; _i++) {
        newStart = hunks[_i].newStart;
        if (newStart > cursorLineNumber) {
          if (nextDiffLineNumber != null) {
            nextDiffLineNumber = Math.min(newStart - 1, nextDiffLineNumber);
          } else {
            nextDiffLineNumber = newStart - 1;
          }
        }
      }
      return this.moveToLineNumber(nextDiffLineNumber);
    };

    GitDiffView.prototype.moveToPreviousDiff = function() {
      var cursorLineNumber, hunks, newStart, previousDiffLineNumber, _i, _len, _ref;
      cursorLineNumber = this.editor.getCursorBufferPosition().row + 1;
      previousDiffLineNumber = -1;
      hunks = (_ref = this.diffs[this.editor.getPath()]) != null ? _ref : [];
      for (_i = 0, _len = hunks.length; _i < _len; _i++) {
        newStart = hunks[_i].newStart;
        if (newStart < cursorLineNumber) {
          previousDiffLineNumber = Math.max(newStart - 1, previousDiffLineNumber);
        }
      }
      return this.moveToLineNumber(previousDiffLineNumber);
    };

    GitDiffView.prototype.moveToLineNumber = function(lineNumber) {
      if (lineNumber == null) {
        lineNumber = -1;
      }
      if (lineNumber >= 0) {
        this.editor.setCursorBufferPosition([lineNumber, 0]);
        return this.editor.moveCursorToFirstCharacterOfLine();
      } else {
        return atom.beep();
      }
    };

    GitDiffView.prototype.unsubscribeFromBuffer = function() {
      if (this.buffer != null) {
        this.removeDiffs();
        if (this.buffer.destroyed) {
          delete this.diffs[this.buffer.getPath()];
        }
        this.buffer.off('contents-modified', this.updateDiffs);
        return this.buffer = null;
      }
    };

    GitDiffView.prototype.subscribeToBuffer = function() {
      this.unsubscribeFromBuffer();
      if (this.buffer = this.editor.getBuffer()) {
        if (this.diffs[this.buffer.getPath()] == null) {
          this.scheduleUpdate();
        }
        return this.buffer.on('contents-modified', this.updateDiffs);
      }
    };

    GitDiffView.prototype.scheduleUpdate = function() {
      return setImmediate(this.updateDiffs);
    };

    GitDiffView.prototype.updateDiffs = function() {
      if (this.buffer == null) {
        return;
      }
      this.generateDiffs();
      return this.renderDiffs();
    };

    GitDiffView.prototype.generateDiffs = function() {
      var path, _ref;
      if (path = this.buffer.getPath()) {
        return this.diffs[path] = (_ref = atom.project.getRepo()) != null ? _ref.getLineDiffs(path, this.buffer.getText()) : void 0;
      }
    };

    GitDiffView.prototype.removeDiffs = function() {
      var klass, _i, _len, _ref;
      if (this.gutter.hasGitLineDiffs) {
        _ref = this.classes;
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          klass = _ref[_i];
          this.gutter.removeClassFromAllLines(klass);
        }
        return this.gutter.hasGitLineDiffs = false;
      }
    };

    GitDiffView.prototype.renderDiffs = function() {
      var hunks, linesHighlighted, newLines, newStart, oldLines, oldStart, row, _i, _j, _k, _len, _ref, _ref1, _ref2, _ref3;
      if (!this.gutter.isVisible()) {
        return;
      }
      this.removeDiffs();
      hunks = (_ref = this.diffs[this.editor.getPath()]) != null ? _ref : [];
      linesHighlighted = false;
      for (_i = 0, _len = hunks.length; _i < _len; _i++) {
        _ref1 = hunks[_i], oldStart = _ref1.oldStart, newStart = _ref1.newStart, oldLines = _ref1.oldLines, newLines = _ref1.newLines;
        if (oldLines === 0 && newLines > 0) {
          for (row = _j = newStart, _ref2 = newStart + newLines; newStart <= _ref2 ? _j < _ref2 : _j > _ref2; row = newStart <= _ref2 ? ++_j : --_j) {
            linesHighlighted |= this.gutter.addClassToLine(row - 1, 'git-line-added');
          }
        } else if (newLines === 0 && oldLines > 0) {
          linesHighlighted |= this.gutter.addClassToLine(newStart - 1, 'git-line-removed');
        } else {
          for (row = _k = newStart, _ref3 = newStart + newLines; newStart <= _ref3 ? _k < _ref3 : _k > _ref3; row = newStart <= _ref3 ? ++_k : --_k) {
            linesHighlighted |= this.gutter.addClassToLine(row - 1, 'git-line-modified');
          }
        }
      }
      return this.gutter.hasGitLineDiffs = linesHighlighted;
    };

    return GitDiffView;

  })();

}).call(this);
