// find-and-replace/lib/project/match-view.js
(function() {
  var LeadingWhitespace, MatchView, Range, View, removeLeadingWhitespace, _ref,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  _ref = require('atom'), View = _ref.View, Range = _ref.Range;

  LeadingWhitespace = /^\s+/;

  removeLeadingWhitespace = function(string) {
    return string.replace(LeadingWhitespace, '');
  };

  module.exports = MatchView = (function(_super) {
    __extends(MatchView, _super);

    function MatchView() {
      this.render = __bind(this.render, this);
      return MatchView.__super__.constructor.apply(this, arguments);
    }

    MatchView.content = function(model, _arg) {
      var filePath, match, matchEnd, matchStart, prefix, range, suffix;
      filePath = _arg.filePath, match = _arg.match;
      range = Range.fromObject(match.range);
      matchStart = range.start.column - match.lineTextOffset;
      matchEnd = range.end.column - match.lineTextOffset;
      prefix = removeLeadingWhitespace(match.lineText.slice(0, matchStart));
      suffix = match.lineText.slice(matchEnd);
      return this.li({
        "class": 'search-result list-item'
      }, (function(_this) {
        return function() {
          _this.span(range.start.row + 1, {
            "class": 'line-number text-subtle'
          });
          return _this.span({
            "class": 'preview'
          }, function() {
            _this.span(prefix);
            _this.span(match.matchText, {
              "class": 'match highlight-info',
              outlet: 'matchText'
            });
            _this.span(match.matchText, {
              "class": 'replacement highlight-success',
              outlet: 'replacementText'
            });
            return _this.span(suffix);
          });
        };
      })(this));
    };

    MatchView.prototype.initialize = function(model, _arg) {
      this.model = model;
      this.filePath = _arg.filePath, this.match = _arg.match;
      this.render();
      return this.subscribe(this.model, 'replacement-pattern-changed', this.render);
    };

    MatchView.prototype.render = function() {
      var replacementText;
      if (this.model.replacementPattern && this.model.regex && (this.model.replacedPathCount == null)) {
        replacementText = this.match.matchText.replace(this.model.regex, this.model.replacementPattern);
        this.replacementText.text(replacementText);
        this.replacementText.show();
        return this.matchText.removeClass('highlight-info').addClass('highlight-error');
      } else {
        this.replacementText.text('').hide();
        return this.matchText.removeClass('highlight-error').addClass('highlight-info');
      }
    };

    MatchView.prototype.confirm = function() {
      return atom.workspaceView.open(this.filePath, {
        split: 'left'
      }).then((function(_this) {
        return function(editor) {
          return editor.setSelectedBufferRange(_this.match.range, {
            autoscroll: true
          });
        };
      })(this));
    };

    return MatchView;

  })(View);

}).call(this);
