// find-and-replace/lib/project/results-view.js
(function() {
  var $, ResultView, ResultsView, ScrollView, _, _ref,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  _ = require('underscore-plus');

  _ref = require('atom'), $ = _ref.$, ScrollView = _ref.ScrollView;

  ResultView = require('./result-view');

  module.exports = ResultsView = (function(_super) {
    __extends(ResultsView, _super);

    function ResultsView() {
      this.clear = __bind(this.clear, this);
      this.removeResult = __bind(this.removeResult, this);
      this.addResult = __bind(this.addResult, this);
      return ResultsView.__super__.constructor.apply(this, arguments);
    }

    ResultsView.content = function() {
      return this.ol({
        "class": 'results-view list-tree focusable-panel',
        tabindex: -1
      });
    };

    ResultsView.prototype.initialize = function(model) {
      this.model = model;
      ResultsView.__super__.initialize.apply(this, arguments);
      this.pixelOverdraw = 100;
      this.lastRenderedResultIndex = 0;
      this.on('core:move-down', (function(_this) {
        return function() {
          return _this.selectNextResult();
        };
      })(this));
      this.on('core:move-up', (function(_this) {
        return function() {
          return _this.selectPreviousResult();
        };
      })(this));
      this.on('scroll resize', (function(_this) {
        return function() {
          if (_this.shouldRenderMoreResults()) {
            return _this.renderResults();
          }
        };
      })(this));
      this.on('core:confirm', (function(_this) {
        return function() {
          var _ref1;
          if ((_ref1 = _this.find('.selected').view()) != null) {
            if (typeof _ref1.confirm === "function") {
              _ref1.confirm();
            }
          }
          return false;
        };
      })(this));
      this.on('mousedown', '.match-result, .path', (function(_this) {
        return function(e) {
          var view;
          _this.find('.selected').removeClass('selected');
          view = $(e.target).view();
          view.addClass('selected');
          return view.confirm();
        };
      })(this));
      this.subscribe(this.model, 'result-added', this.addResult);
      this.subscribe(this.model, 'result-removed', this.removeResult);
      this.subscribe(this.model, 'search-state-cleared', this.clear);
      return this.renderResults();
    };

    ResultsView.prototype.beforeRemove = function() {
      return this.clear();
    };

    ResultsView.prototype.hasResults = function() {
      return this.model.getResultCount() > 0;
    };

    ResultsView.prototype.addResult = function(filePath, result) {
      var resultView;
      resultView = this.getResultView(filePath);
      if (resultView) {
        return resultView.renderResult(result);
      } else {
        this.renderResults();
        if (this.getPathCount() === 1) {
          return this.selectFirstResult();
        }
      }
    };

    ResultsView.prototype.removeResult = function(filePath) {
      var resultView;
      resultView = this.getResultView(filePath);
      if (resultView) {
        return resultView.renderResult(null);
      }
    };

    ResultsView.prototype.renderResults = function(_arg) {
      var filePath, paths, renderAll, result, resultView, _i, _len, _ref1;
      renderAll = (_arg != null ? _arg : {}).renderAll;
      if (!(renderAll || this.shouldRenderMoreResults())) {
        return;
      }
      paths = this.model.getPaths();
      _ref1 = paths.slice(this.lastRenderedResultIndex);
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        filePath = _ref1[_i];
        result = this.model.getResult(filePath);
        if (!renderAll && !this.shouldRenderMoreResults()) {
          break;
        }
        resultView = new ResultView(this.model, filePath, result);
        this.append(resultView);
        this.lastRenderedResultIndex++;
      }
      return null;
    };

    ResultsView.prototype.shouldRenderMoreResults = function() {
      return this.prop('scrollHeight') <= this.height() + this.pixelOverdraw || this.scrollBottom() + this.pixelOverdraw >= this.prop('scrollHeight');
    };

    ResultsView.prototype.selectFirstResult = function() {
      return this.find('.search-result:first').addClass('selected');
    };

    ResultsView.prototype.selectNextResult = function() {
      var nextView, selectedView;
      selectedView = this.find('.selected').view();
      if (!selectedView) {
        return this.selectFirstResult();
      }
      if (selectedView instanceof ResultView) {
        nextView = selectedView.find('.search-result:first').view();
      } else {
        nextView = selectedView.next().view();
        if (nextView == null) {
          nextView = selectedView.closest('.path').next().find('.search-result:first').view();
        }
      }
      if (nextView != null) {
        selectedView.removeClass('selected');
        nextView.addClass('selected');
        return this.scrollTo(nextView);
      }
    };

    ResultsView.prototype.selectPreviousResult = function() {
      var previousView, selectedView;
      selectedView = this.find('.selected').view();
      if (!selectedView) {
        return this.selectFirstResult();
      }
      if (selectedView instanceof ResultView) {
        previousView = selectedView.prev().find('.search-result:last').view();
      } else {
        previousView = selectedView.prev().view();
        if (previousView == null) {
          previousView = selectedView.closest('.path').prev().find('.search-result:last').view();
        }
      }
      if (previousView != null) {
        selectedView.removeClass('selected');
        previousView.addClass('selected');
        return this.scrollTo(previousView);
      }
    };

    ResultsView.prototype.getPathCount = function() {
      return this.model.getPathCount();
    };

    ResultsView.prototype.getMatchCount = function() {
      return this.model.getMatchCount();
    };

    ResultsView.prototype.clear = function() {
      this.lastRenderedResultIndex = 0;
      return this.empty();
    };

    ResultsView.prototype.scrollTo = function(element) {
      var bottom, top;
      top = this.scrollTop() + element.offset().top - this.offset().top;
      bottom = top + element.outerHeight();
      if (bottom > this.scrollBottom()) {
        this.scrollBottom(bottom);
      }
      if (top < this.scrollTop()) {
        return this.scrollTop(top);
      }
    };

    ResultsView.prototype.scrollToBottom = function() {
      var lastPath;
      this.renderResults({
        renderAll: true
      });
      ResultsView.__super__.scrollToBottom.call(this);
      this.find('.selected').removeClass('selected');
      lastPath = this.find('.path:last');
      return lastPath.find('.search-result:last').addClass('selected');
    };

    ResultsView.prototype.scrollToTop = function() {
      ResultsView.__super__.scrollToTop.call(this);
      this.find('.selected').removeClass('selected');
      return this.find('.path:first').addClass('selected');
    };

    ResultsView.prototype.getResultView = function(filePath) {
      var el;
      el = this.find("[data-path=\"" + (_.escapeAttribute(filePath)) + "\"]");
      if (el.length) {
        return el.view();
      } else {
        return null;
      }
    };

    return ResultsView;

  })(ScrollView);

}).call(this);
