// find-and-replace/lib/project/results-model.js
(function() {
  var Emitter, Q, Result, ResultsModel, escapeHelper, _,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  Q = require('q');

  _ = require('underscore-plus');

  Emitter = require('emissary').Emitter;

  escapeHelper = require('../escape-helper');

  Result = (function() {
    Result.create = function(result) {
      var _ref;
      if (result != null ? (_ref = result.matches) != null ? _ref.length : void 0 : void 0) {
        return new Result(result);
      } else {
        return null;
      }
    };

    function Result(result) {
      _.extend(this, result);
    }

    return Result;

  })();

  module.exports = ResultsModel = (function() {
    Emitter.includeInto(ResultsModel);

    function ResultsModel(state) {
      var _ref, _ref1;
      if (state == null) {
        state = {};
      }
      this.onContentsModified = __bind(this.onContentsModified, this);
      this.useRegex = (_ref = state.useRegex) != null ? _ref : false;
      this.caseSensitive = (_ref1 = state.caseSensitive) != null ? _ref1 : false;
      atom.project.eachEditor((function(_this) {
        return function(editSession) {
          return editSession.on('contents-modified', function() {
            return _this.onContentsModified(editSession);
          });
        };
      })(this));
      this.clear();
    }

    ResultsModel.prototype.serialize = function() {
      return {
        useRegex: this.useRegex,
        caseSensitive: this.caseSensitive
      };
    };

    ResultsModel.prototype.clear = function() {
      this.clearSearchState();
      this.clearReplacementState();
      return this.emit('cleared', this.getResultsSummary());
    };

    ResultsModel.prototype.clearSearchState = function() {
      this.pathCount = 0;
      this.matchCount = 0;
      this.regex = null;
      this.results = {};
      this.paths = [];
      this.active = false;
      this.pattern = '';
      if (this.inProgressSearchPromise != null) {
        this.inProgressSearchPromise.cancel();
        this.inProgressSearchPromise = null;
      }
      return this.emit('search-state-cleared', this.getResultsSummary());
    };

    ResultsModel.prototype.clearReplacementState = function() {
      this.replacementPattern = null;
      this.replacedPathCount = null;
      this.replacementCount = null;
      return this.emit('replacement-state-cleared', this.getResultsSummary());
    };

    ResultsModel.prototype.search = function(pattern, searchPaths, replacementPattern, _arg) {
      var keepReplacementState, onPathsSearched, onlyRunIfChanged, _ref;
      _ref = _arg != null ? _arg : {}, onlyRunIfChanged = _ref.onlyRunIfChanged, keepReplacementState = _ref.keepReplacementState;
      if (onlyRunIfChanged && (pattern != null) && (searchPaths != null) && pattern === this.pattern && _.isEqual(searchPaths, this.searchedPaths)) {
        return Q();
      }
      if (keepReplacementState) {
        this.clearSearchState();
      } else {
        this.clear();
      }
      this.active = true;
      this.regex = this.getRegex(pattern);
      this.pattern = pattern;
      this.searchedPaths = searchPaths;
      this.updateReplacementPattern(replacementPattern);
      onPathsSearched = (function(_this) {
        return function(numberOfPathsSearched) {
          return _this.emit('paths-searched', numberOfPathsSearched);
        };
      })(this);
      this.inProgressSearchPromise = atom.project.scan(this.regex, {
        paths: searchPaths,
        onPathsSearched: onPathsSearched
      }, (function(_this) {
        return function(result) {
          return _this.setResult(result.filePath, Result.create(result));
        };
      })(this));
      this.emit('search', this.inProgressSearchPromise);
      return this.inProgressSearchPromise.then((function(_this) {
        return function(message) {
          if (message === 'cancelled') {
            return _this.emit('cancelled-searching');
          } else {
            _this.inProgressSearchPromise = null;
            return _this.emit('finished-searching', _this.getResultsSummary());
          }
        };
      })(this));
    };

    ResultsModel.prototype.replace = function(pattern, searchPaths, replacementPattern, replacementPaths) {
      var promise, regex;
      regex = this.getRegex(pattern);
      this.updateReplacementPattern(replacementPattern);
      replacementPattern = escapeHelper.unescapeEscapeSequence(replacementPattern);
      this.active = false;
      this.replacedPathCount = 0;
      this.replacementCount = 0;
      promise = atom.project.replace(regex, replacementPattern, replacementPaths, (function(_this) {
        return function(result) {
          if (result && result.replacements) {
            _this.replacedPathCount++;
            _this.replacementCount += result.replacements;
          }
          return _this.emit('path-replaced', result);
        };
      })(this));
      this.emit('replace', promise);
      return promise.then((function(_this) {
        return function() {
          _this.emit('finished-replacing', _this.getResultsSummary());
          return _this.search(pattern, searchPaths, replacementPattern, {
            keepReplacementState: true
          });
        };
      })(this));
    };

    ResultsModel.prototype.updateReplacementPattern = function(replacementPattern) {
      this.replacementPattern = replacementPattern || null;
      return this.emit('replacement-pattern-changed', this.regex, replacementPattern);
    };

    ResultsModel.prototype.setActive = function(isActive) {
      if ((isActive && this.pattern) || !isActive) {
        return this.active = isActive;
      }
    };

    ResultsModel.prototype.getActive = function() {
      return this.active;
    };

    ResultsModel.prototype.toggleUseRegex = function() {
      return this.useRegex = !this.useRegex;
    };

    ResultsModel.prototype.toggleCaseSensitive = function() {
      return this.caseSensitive = !this.caseSensitive;
    };

    ResultsModel.prototype.getResultsSummary = function() {
      var pattern;
      pattern = this.pattern || '';
      return {
        pattern: pattern,
        pathCount: this.pathCount,
        matchCount: this.matchCount,
        replacementPattern: this.replacementPattern,
        replacedPathCount: this.replacedPathCount,
        replacementCount: this.replacementCount
      };
    };

    ResultsModel.prototype.getPathCount = function() {
      return this.pathCount;
    };

    ResultsModel.prototype.getMatchCount = function() {
      return this.matchCount;
    };

    ResultsModel.prototype.getPattern = function() {
      return this.pattern || '';
    };

    ResultsModel.prototype.getPaths = function() {
      return this.paths;
    };

    ResultsModel.prototype.getResult = function(filePath) {
      return this.results[filePath];
    };

    ResultsModel.prototype.setResult = function(filePath, result) {
      if (result) {
        return this.addResult(filePath, result);
      } else {
        return this.removeResult(filePath);
      }
    };

    ResultsModel.prototype.addResult = function(filePath, result) {
      if (this.results[filePath]) {
        this.matchCount -= this.results[filePath].matches.length;
      } else {
        this.pathCount++;
        this.paths.push(filePath);
      }
      this.matchCount += result.matches.length;
      this.results[filePath] = result;
      return this.emit('result-added', filePath, result);
    };

    ResultsModel.prototype.removeResult = function(filePath) {
      if (this.results[filePath]) {
        this.pathCount--;
        this.matchCount -= this.results[filePath].matches.length;
        this.paths = _.without(this.paths, filePath);
        delete this.results[filePath];
        return this.emit('result-removed', filePath);
      }
    };

    ResultsModel.prototype.getRegex = function(pattern) {
      var flags;
      flags = 'g';
      if (!this.caseSensitive) {
        flags += 'i';
      }
      pattern = escapeHelper.unescapeEscapeSequence(pattern);
      if (this.useRegex) {
        return new RegExp(pattern, flags);
      } else {
        return new RegExp(_.escapeRegExp(pattern), flags);
      }
    };

    ResultsModel.prototype.onContentsModified = function(editSession) {
      var matches, result;
      if (!this.active) {
        return;
      }
      if (!editSession.getPath()) {
        return;
      }
      matches = [];
      editSession.scan(this.regex, function(match) {
        return matches.push(match);
      });
      result = Result.create({
        matches: matches
      });
      this.setResult(editSession.getPath(), result);
      return this.emit('finished-searching', this.getResultsSummary());
    };

    return ResultsModel;

  })();

}).call(this);
