// find-and-replace/lib/project/results-pane.js
(function() {
  var ResultsPaneView, ResultsView, ScrollView, Util,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  ScrollView = require('atom').ScrollView;

  ResultsView = require('./results-view');

  Util = require('./util');

  module.exports = ResultsPaneView = (function(_super) {
    __extends(ResultsPaneView, _super);

    function ResultsPaneView() {
      this.onCleared = __bind(this.onCleared, this);
      this.onReplacementStateCleared = __bind(this.onReplacementStateCleared, this);
      this.onFinishedSearching = __bind(this.onFinishedSearching, this);
      this.onPathsSearched = __bind(this.onPathsSearched, this);
      this.onSearch = __bind(this.onSearch, this);
      return ResultsPaneView.__super__.constructor.apply(this, arguments);
    }

    ResultsPaneView.URI = "atom://find-and-replace/project-results";

    ResultsPaneView.content = function() {
      return this.div({
        "class": 'preview-pane pane-item'
      }, (function(_this) {
        return function() {
          _this.div({
            "class": 'panel-heading'
          }, function() {
            _this.span({
              outlet: 'previewCount',
              "class": 'preview-count inline-block'
            });
            return _this.div({
              outlet: 'loadingMessage',
              "class": 'inline-block'
            }, function() {
              _this.div({
                "class": 'loading loading-spinner-tiny inline-block'
              });
              return _this.div({
                outlet: 'searchedCountBlock',
                "class": 'inline-block'
              }, function() {
                _this.span({
                  outlet: 'searchedCount',
                  "class": 'searched-count'
                });
                return _this.span(' paths searched');
              });
            });
          });
          _this.subview('resultsView', new ResultsView(_this.model));
          return _this.ul({
            "class": 'centered background-message no-results-overlay'
          }, function() {
            return _this.li('No Results');
          });
        };
      })(this));
    };

    ResultsPaneView.prototype.initialize = function() {
      ResultsPaneView.__super__.initialize.apply(this, arguments);
      this.loadingMessage.hide();
      this.model = this.constructor.model;
      this.model.setActive(true);
      this.handleEvents();
      return this.onFinishedSearching(this.model.getResultsSummary());
    };

    ResultsPaneView.prototype.beforeRemove = function() {
      return this.model.setActive(false);
    };

    ResultsPaneView.prototype.copy = function() {
      return new ResultsPaneView();
    };

    ResultsPaneView.prototype.getPane = function() {
      return this.parents('.pane').view();
    };

    ResultsPaneView.prototype.getTitle = function() {
      return "Project Find Results";
    };

    ResultsPaneView.prototype.getUri = function() {
      return this.constructor.URI;
    };

    ResultsPaneView.prototype.focus = function() {
      return this.resultsView.focus();
    };

    ResultsPaneView.prototype.handleEvents = function() {
      this.subscribe(this.model, 'search', this.onSearch);
      this.subscribe(this.model, 'cleared', this.onCleared);
      this.subscribe(this.model, 'replacement-state-cleared', this.onReplacementStateCleared);
      this.subscribe(this.model, 'finished-searching', this.onFinishedSearching);
      return this.subscribe(this.model, 'paths-searched', this.onPathsSearched);
    };

    ResultsPaneView.prototype.onSearch = function(deferred) {
      var timeout;
      this.loadingMessage.show();
      this.previewCount.text('Searching...');
      this.searchedCount.text('0');
      this.searchedCountBlock.hide();
      this.removeClass('no-results');
      this.previewCount.show();
      this.showSearchedCountBlock = false;
      timeout = setTimeout((function(_this) {
        return function() {
          _this.searchedCountBlock.show();
          return _this.showSearchedCountBlock = true;
        };
      })(this), 500);
      return deferred.done((function(_this) {
        return function() {
          return _this.loadingMessage.hide();
        };
      })(this));
    };

    ResultsPaneView.prototype.onPathsSearched = function(numberOfPathsSearched) {
      if (this.showSearchedCountBlock) {
        return this.searchedCount.text(numberOfPathsSearched);
      }
    };

    ResultsPaneView.prototype.onFinishedSearching = function(results) {
      this.hideOrShowNoResults(results);
      return this.previewCount.html(Util.getSearchResultsMessage(results));
    };

    ResultsPaneView.prototype.onReplacementStateCleared = function(results) {
      this.hideOrShowNoResults(results);
      return this.previewCount.html(Util.getSearchResultsMessage(results));
    };

    ResultsPaneView.prototype.onCleared = function() {
      this.addClass('no-results');
      this.previewCount.text('Find in project results');
      this.loadingMessage.hide();
      return this.searchedCountBlock.hide();
    };

    ResultsPaneView.prototype.hideOrShowNoResults = function(results) {
      if (results.pathCount) {
        return this.removeClass('no-results');
      } else {
        return this.addClass('no-results');
      }
    };

    return ResultsPaneView;

  })(ScrollView);

}).call(this);
