// metrics/lib/reporter.js
(function() {
  var Reporter, https, path, querystring, _;

  https = require('https');

  path = require('path');

  querystring = require('querystring');

  _ = require('underscore-plus');

  module.exports = Reporter = (function() {
    function Reporter() {}

    Reporter.sendEvent = function(category, name, value) {
      var params;
      params = {
        t: 'event',
        ec: category,
        ea: name,
        ev: value
      };
      return this.send(params);
    };

    Reporter.sendTiming = function(category, name, value) {
      var params;
      params = {
        t: 'timing',
        utc: category,
        utv: name,
        utt: value
      };
      return this.send(params);
    };

    Reporter.viewNameForPaneItem = function(item) {
      var extension, itemPath, name, _ref;
      name = (_ref = typeof item.getViewClass === "function" ? item.getViewClass().name : void 0) != null ? _ref : item.constructor.name;
      if (name !== 'EditorView') {
        return name;
      }
      itemPath = typeof item.getPath === "function" ? item.getPath() : void 0;
      if (path.dirname(itemPath) !== atom.getConfigDirPath()) {
        return name;
      }
      extension = path.extname(itemPath);
      switch (path.basename(itemPath, extension)) {
        case 'config':
          if (extension === '.json' || extension === '.cson') {
            name = 'UserConfig';
          }
          break;
        case 'init':
          if (extension === '.js' || extension === '.coffee') {
            name = 'UserInitScript';
          }
          break;
        case 'keymap':
          if (extension === '.json' || extension === '.cson') {
            name = 'UserKeymap';
          }
          break;
        case 'snippets':
          if (extension === '.json' || extension === '.cson') {
            name = 'UserSnippets';
          }
          break;
        case 'styles':
          if (extension === '.css' || extension === '.less') {
            name = 'UserStylesheet';
          }
      }
      return name;
    };

    Reporter.sendPaneItem = function(item) {
      var params;
      params = {
        t: 'appview',
        cd: this.viewNameForPaneItem(item)
      };
      return this.send(params);
    };

    Reporter.send = function(params) {
      _.extend(params, this.defaultParams());
      return this.request({
        method: 'POST',
        hostname: 'www.google-analytics.com',
        path: "/collect?" + (querystring.stringify(params)),
        headers: {
          'User-Agent': navigator.userAgent
        }
      });
    };

    Reporter.request = function(options) {
      var request;
      request = https.request(options);
      request.on('error', function() {});
      return request.end();
    };

    Reporter.defaultParams = function() {
      return {
        v: 1,
        tid: "UA-3769691-33",
        cid: atom.config.get('metrics.userId'),
        an: 'atom',
        av: atom.getVersion(),
        sr: "" + screen.width + "x" + screen.height
      };
    };

    return Reporter;

  })();

}).call(this);
