// bracket-matcher/lib/bracket-matcher-view.js
(function() {
  var BracketMatcherView, Range, View, endPairMatches, startPairMatches, _, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  _ = require('underscore-plus');

  _ref = require('atom'), Range = _ref.Range, View = _ref.View;

  startPairMatches = {
    '(': ')',
    '[': ']',
    '{': '}'
  };

  endPairMatches = {
    ')': '(',
    ']': '[',
    '}': '{'
  };

  module.exports = BracketMatcherView = (function(_super) {
    __extends(BracketMatcherView, _super);

    function BracketMatcherView() {
      return BracketMatcherView.__super__.constructor.apply(this, arguments);
    }

    BracketMatcherView.content = function() {
      return this.div((function(_this) {
        return function() {
          _this.div({
            "class": 'bracket-matcher',
            style: 'display: none',
            outlet: 'startView'
          });
          return _this.div({
            "class": 'bracket-matcher',
            style: 'display: none',
            outlet: 'endView'
          });
        };
      })(this));
    };

    BracketMatcherView.prototype.initialize = function(editorView) {
      this.editorView = editorView;
      this.editor = this.editorView.editor;
      this.pairHighlighted = false;
      this.bufferChanged = false;
      this.subscribe(atom.config.observe('editor.fontSize', (function(_this) {
        return function() {
          return _this.updateMatch();
        };
      })(this)));
      this.subscribe(this.editor.getBuffer(), 'changed', (function(_this) {
        return function() {
          return _this.bufferChanged = true;
        };
      })(this));
      this.subscribe(this.editorView, 'editor:display-updated', (function(_this) {
        return function() {
          if (_this.bufferChanged) {
            _this.bufferChanged = false;
            return _this.updateMatch();
          }
        };
      })(this));
      this.subscribe(this.editor.getCursor(), 'moved', (function(_this) {
        return function() {
          return _this.updateMatch();
        };
      })(this));
      this.subscribeToCommand(this.editorView, 'bracket-matcher:go-to-matching-bracket', (function(_this) {
        return function() {
          return _this.goToMatchingPair();
        };
      })(this));
      this.subscribeToCommand(this.editorView, 'bracket-matcher:go-to-enclosing-bracket', (function(_this) {
        return function() {
          return _this.goToEnclosingPair();
        };
      })(this));
      this.editorView.underlayer.append(this);
      return this.updateMatch();
    };

    BracketMatcherView.prototype.updateMatch = function() {
      var currentPair, matchPosition, matchingPair, position, _ref1, _ref2;
      if (this.pairHighlighted) {
        this.startView.hide();
        this.endView.hide();
      }
      this.pairHighlighted = false;
      if (!this.editor.getSelection().isEmpty()) {
        return;
      }
      if (this.editor.isFoldedAtCursorRow()) {
        return;
      }
      _ref1 = this.findCurrentPair(startPairMatches), position = _ref1.position, currentPair = _ref1.currentPair, matchingPair = _ref1.matchingPair;
      if (position) {
        matchPosition = this.findMatchingEndPair(position, currentPair, matchingPair);
      } else {
        _ref2 = this.findCurrentPair(endPairMatches), position = _ref2.position, currentPair = _ref2.currentPair, matchingPair = _ref2.matchingPair;
        if (position) {
          matchPosition = this.findMatchingStartPair(position, matchingPair, currentPair);
        }
      }
      if ((position != null) && (matchPosition != null)) {
        this.moveHighlightViews([position, matchPosition]);
        return this.pairHighlighted = true;
      }
    };

    BracketMatcherView.prototype.findMatchingEndPair = function(startPairPosition, startPair, endPair) {
      var endPairPosition, regex, scanRange, unpairedCount;
      scanRange = new Range(startPairPosition.translate([0, 1]), this.editor.buffer.getEndPosition());
      regex = new RegExp("[" + (_.escapeRegExp(startPair + endPair)) + "]", 'g');
      endPairPosition = null;
      unpairedCount = 0;
      this.editor.scanInBufferRange(regex, scanRange, (function(_this) {
        return function(_arg) {
          var match, range, stop;
          match = _arg.match, range = _arg.range, stop = _arg.stop;
          if (match[0] === startPair) {
            return unpairedCount++;
          } else if (match[0] === endPair) {
            unpairedCount--;
            endPairPosition = range.start;
            if (unpairedCount < 0) {
              return stop();
            }
          }
        };
      })(this));
      return endPairPosition;
    };

    BracketMatcherView.prototype.findMatchingStartPair = function(endPairPosition, startPair, endPair) {
      var regex, scanRange, startPairPosition, unpairedCount;
      scanRange = new Range([0, 0], endPairPosition);
      regex = new RegExp("[" + (_.escapeRegExp(startPair + endPair)) + "]", 'g');
      startPairPosition = null;
      unpairedCount = 0;
      this.editor.backwardsScanInBufferRange(regex, scanRange, (function(_this) {
        return function(_arg) {
          var match, range, stop;
          match = _arg.match, range = _arg.range, stop = _arg.stop;
          if (match[0] === endPair) {
            return unpairedCount++;
          } else if (match[0] === startPair) {
            unpairedCount--;
            startPairPosition = range.start;
            if (unpairedCount < 0) {
              return stop();
            }
          }
        };
      })(this));
      return startPairPosition;
    };

    BracketMatcherView.prototype.findAnyStartPair = function(cursorPosition) {
      var combinedRegExp, endPair, endPairRegExp, scanRange, startPair, startPairRegExp, startPosition, unpairedCount;
      scanRange = new Range([0, 0], cursorPosition);
      startPair = _.escapeRegExp(_.keys(startPairMatches).join(''));
      endPair = _.escapeRegExp(_.keys(endPairMatches).join(''));
      combinedRegExp = new RegExp("[" + startPair + endPair + "]", 'g');
      startPairRegExp = new RegExp("[" + startPair + "]", 'g');
      endPairRegExp = new RegExp("[" + endPair + "]", 'g');
      startPosition = null;
      unpairedCount = 0;
      this.editor.backwardsScanInBufferRange(combinedRegExp, scanRange, (function(_this) {
        return function(_arg) {
          var match, range, stop;
          match = _arg.match, range = _arg.range, stop = _arg.stop;
          if (match[0].match(endPairRegExp)) {
            return unpairedCount++;
          } else if (match[0].match(startPairRegExp)) {
            unpairedCount--;
            startPosition = range.start;
            if (unpairedCount < 0) {
              return stop();
            }
          }
        };
      })(this));
      return startPosition;
    };

    BracketMatcherView.prototype.moveHighlightView = function(view, bufferPosition, pixelPosition) {
      var element;
      view.bufferPosition = bufferPosition;
      element = view[0];
      element.style.display = 'block';
      element.style.top = "" + pixelPosition.top + "px";
      element.style.left = "" + pixelPosition.left + "px";
      element.style.width = "" + this.editorView.charWidth + "px";
      return element.style.height = "" + this.editorView.lineHeight + "px";
    };

    BracketMatcherView.prototype.moveHighlightViews = function(bufferRange) {
      var end, endPixelPosition, start, startPixelPosition, _ref1;
      _ref1 = Range.fromObject(bufferRange), start = _ref1.start, end = _ref1.end;
      startPixelPosition = this.editorView.pixelPositionForBufferPosition(start);
      endPixelPosition = this.editorView.pixelPositionForBufferPosition(end);
      this.moveHighlightView(this.startView, start, startPixelPosition);
      return this.moveHighlightView(this.endView, end, endPixelPosition);
    };

    BracketMatcherView.prototype.findCurrentPair = function(matches) {
      var currentPair, matchingPair, position;
      position = this.editor.getCursorBufferPosition();
      currentPair = this.editor.getTextInRange(Range.fromPointWithDelta(position, 0, 1));
      if (!matches[currentPair]) {
        position = position.translate([0, -1]);
        currentPair = this.editor.getTextInRange(Range.fromPointWithDelta(position, 0, 1));
      }
      if (matchingPair = matches[currentPair]) {
        return {
          position: position,
          currentPair: currentPair,
          matchingPair: matchingPair
        };
      } else {
        return {};
      }
    };

    BracketMatcherView.prototype.goToMatchingPair = function() {
      var endPosition, position, previousPosition, startPosition;
      if (!this.pairHighlighted) {
        return this.goToEnclosingPair();
      }
      if (!this.editorView.underlayer.isVisible()) {
        return;
      }
      position = this.editor.getCursorBufferPosition();
      previousPosition = position.translate([0, -1]);
      startPosition = this.startView.bufferPosition;
      endPosition = this.endView.bufferPosition;
      if (position.isEqual(startPosition)) {
        return this.editor.setCursorBufferPosition(endPosition.translate([0, 1]));
      } else if (previousPosition.isEqual(startPosition)) {
        return this.editor.setCursorBufferPosition(endPosition);
      } else if (position.isEqual(endPosition)) {
        return this.editor.setCursorBufferPosition(startPosition.translate([0, 1]));
      } else if (previousPosition.isEqual(endPosition)) {
        return this.editor.setCursorBufferPosition(startPosition);
      }
    };

    BracketMatcherView.prototype.goToEnclosingPair = function() {
      var matchPosition, position;
      if (this.pairHighlighted) {
        return;
      }
      if (!this.editorView.underlayer.isVisible()) {
        return;
      }
      position = this.editor.getCursorBufferPosition();
      matchPosition = this.findAnyStartPair(position);
      if (matchPosition) {
        return this.editor.setCursorBufferPosition(matchPosition);
      }
    };

    return BracketMatcherView;

  })(View);

}).call(this);
