// status-bar/lib/status-bar.js
(function() {
  var CursorPositionView, FileInfoView, GitView, StatusBarView;

  StatusBarView = require('./status-bar-view');

  FileInfoView = require('./file-info-view');

  CursorPositionView = require('./cursor-position-view');

  GitView = require('./git-view');

  module.exports = {
    activate: function() {
      var DevModeView;
      this.statusBar = new StatusBarView();
      this.statusBar.attach();
      if (atom.getLoadSettings().devMode) {
        DevModeView = require('./dev-mode-view');
        this.statusBar.appendLeft(new DevModeView());
      }
      this.fileInfo = new FileInfoView(this.statusBar);
      this.statusBar.appendLeft(this.fileInfo);
      this.cursorPosition = new CursorPositionView(this.statusBar);
      this.statusBar.appendLeft(this.cursorPosition);
      this.git = new GitView(this.statusBar);
      return this.statusBar.appendRight(this.git);
    },
    deactivate: function() {
      var _ref, _ref1, _ref2, _ref3;
      if ((_ref = this.git) != null) {
        _ref.destroy();
      }
      this.git = null;
      if ((_ref1 = this.fileInfo) != null) {
        _ref1.destroy();
      }
      this.fileInfo = null;
      if ((_ref2 = this.cursorPosition) != null) {
        _ref2.destroy();
      }
      this.cursorPosition = null;
      if ((_ref3 = this.statusBar) != null) {
        _ref3.destroy();
      }
      return this.statusBar = null;
    }
  };

}).call(this);
