// wrap-guide/lib/wrap-guide-view.js
(function() {
  var $, View, WrapGuideView, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  _ref = require('atom'), $ = _ref.$, View = _ref.View;

  module.exports = WrapGuideView = (function(_super) {
    __extends(WrapGuideView, _super);

    function WrapGuideView() {
      return WrapGuideView.__super__.constructor.apply(this, arguments);
    }

    WrapGuideView.activate = function() {
      return atom.workspaceView.eachEditorView(function(editorView) {
        if (editorView.attached && editorView.getPane()) {
          return editorView.underlayer.append(new WrapGuideView(editorView));
        }
      });
    };

    WrapGuideView.content = function() {
      return this.div({
        "class": 'wrap-guide'
      });
    };

    WrapGuideView.prototype.initialize = function(editorView) {
      this.editorView = editorView;
      this.subscribe(atom.config.observe('editor.fontSize', {
        callNow: false
      }, (function(_this) {
        return function() {
          return _this.updateGuide();
        };
      })(this)));
      this.subscribe(atom.config.observe('editor.preferredLineLength', {
        callNow: false
      }, (function(_this) {
        return function() {
          return _this.updateGuide();
        };
      })(this)));
      this.subscribe(atom.config.observe('wrap-guide.columns', {
        callNow: false
      }, (function(_this) {
        return function() {
          return _this.updateGuide();
        };
      })(this)));
      this.subscribe(this.editorView, 'editor:path-changed', (function(_this) {
        return function() {
          return _this.updateGuide();
        };
      })(this));
      this.subscribe(this.editorView, 'editor:min-width-changed', (function(_this) {
        return function() {
          return _this.updateGuide();
        };
      })(this));
      this.subscribe($(window), 'resize', (function(_this) {
        return function() {
          return _this.updateGuide();
        };
      })(this));
      return this.updateGuide();
    };

    WrapGuideView.prototype.getDefaultColumn = function() {
      return atom.config.getPositiveInt('editor.preferredLineLength', 80);
    };

    WrapGuideView.prototype.getGuideColumn = function(path) {
      var column, customColumn, customColumns, pattern, regex, _i, _len;
      customColumns = atom.config.get('wrap-guide.columns');
      if (!Array.isArray(customColumns)) {
        return this.getDefaultColumn();
      }
      for (_i = 0, _len = customColumns.length; _i < _len; _i++) {
        customColumn = customColumns[_i];
        if (!(typeof customColumn === 'object')) {
          continue;
        }
        pattern = customColumn.pattern, column = customColumn.column;
        if (!pattern) {
          continue;
        }
        try {
          regex = new RegExp(pattern);
        } catch (_error) {
          continue;
        }
        if (regex.test(path)) {
          return parseInt(column);
        }
      }
      return this.getDefaultColumn();
    };

    WrapGuideView.prototype.updateGuide = function() {
      var column, columnWidth;
      column = this.getGuideColumn(this.editorView.getEditor().getPath());
      if (column > 0) {
        columnWidth = this.editorView.charWidth * column;
        if (columnWidth < this.editorView.layerMinWidth || columnWidth < this.editorView.width()) {
          return this.css('left', columnWidth).show();
        } else {
          return this.hide();
        }
      } else {
        return this.hide();
      }
    };

    return WrapGuideView;

  })(View);

}).call(this);
