// settings-view/lib/settings-view.js
(function() {
  var $, $$, CSON, EditorView, GeneralPanel, InstalledPackageView, KeybindingsPanel, PackageManager, PackageMenuView, PackagesPanel, ScrollView, SettingsView, ThemesPanel, async, fuzzaldrin, path, _, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  path = require('path');

  _ = require('underscore-plus');

  _ref = require('atom'), $ = _ref.$, $$ = _ref.$$, ScrollView = _ref.ScrollView, EditorView = _ref.EditorView;

  async = require('async');

  CSON = require('season');

  fuzzaldrin = require('fuzzaldrin');

  GeneralPanel = require('./general-panel');

  InstalledPackageView = require('./installed-package-view');

  KeybindingsPanel = require('./keybindings-panel');

  PackageManager = require('./package-manager');

  PackageMenuView = require('./package-menu-view');

  PackagesPanel = require('./packages-panel');

  ThemesPanel = require('./themes-panel');

  module.exports = SettingsView = (function(_super) {
    __extends(SettingsView, _super);

    function SettingsView() {
      return SettingsView.__super__.constructor.apply(this, arguments);
    }

    SettingsView.content = function() {
      return this.div({
        "class": 'settings-view pane-item',
        tabindex: -1
      }, (function(_this) {
        return function() {
          _this.div({
            "class": 'config-menu',
            outlet: 'sidebar'
          }, function() {
            _this.div({
              "class": 'atom-banner'
            });
            _this.ul({
              "class": 'panels-menu nav nav-pills nav-stacked',
              outlet: 'panelMenu'
            }, function() {
              _this.div({
                "class": 'panel-menu-separator',
                outlet: 'menuSeparator'
              });
              return _this.div({
                "class": 'editor-container settings-filter'
              }, function() {
                return _this.subview('filterEditor', new EditorView({
                  mini: true,
                  placeholderText: 'Filter packages'
                }));
              });
            });
            _this.ul({
              "class": 'panels-packages nav nav-pills nav-stacked',
              outlet: 'panelPackages'
            });
            return _this.div({
              "class": 'button-area'
            }, function() {
              return _this.button({
                "class": 'btn btn-default icon icon-link-external',
                outlet: 'openDotAtom'
              }, 'Open ~/.atom');
            });
          });
          return _this.div({
            "class": 'panels padded',
            outlet: 'panels'
          });
        };
      })(this));
    };

    SettingsView.prototype.initialize = function(_arg) {
      var activePanelName, _ref1;
      _ref1 = _arg != null ? _arg : {}, this.uri = _ref1.uri, activePanelName = _ref1.activePanelName;
      SettingsView.__super__.initialize.apply(this, arguments);
      this.packageManager = new PackageManager();
      this.handlePackageEvents();
      this.panelToShow = activePanelName;
      this.filterEditor.hide();
      return process.nextTick((function(_this) {
        return function() {
          return _this.activatePackages(function() {
            return _this.initializePanels();
          });
        };
      })(this));
    };

    SettingsView.prototype.handlePackageEvents = function() {
      this.subscribe(this.packageManager, 'package-installed theme-installed', (function(_this) {
        return function(_arg) {
          var addedPackageElement, beforeElement, compare, name, pack, panelMenuItem, title, _i, _len, _ref1;
          name = _arg.name;
          if (pack = atom.packages.getLoadedPackage(name)) {
            title = _this.packageManager.getPackageTitle(pack);
            _this.addPackagePanel(pack);
            _ref1 = _this.panelPackages.children();
            for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
              panelMenuItem = _ref1[_i];
              compare = title.localeCompare($(panelMenuItem).view().nameLabel.text());
              if (compare > 0) {
                beforeElement = panelMenuItem;
              } else if (compare === 0) {
                addedPackageElement = panelMenuItem;
              }
            }
            if ((beforeElement != null) && (addedPackageElement != null)) {
              return $(addedPackageElement).insertAfter(beforeElement);
            }
          }
        };
      })(this));
      return this.subscribe(this.packageManager, 'package-uninstalled theme-uninstalled', (function(_this) {
        return function(_arg) {
          var name;
          name = _arg.name;
          _this.removePanel(name);
          if (name === _this.activePanelName) {
            return _this.showPanel('Packages');
          }
        };
      })(this));
    };

    SettingsView.prototype.initializePanels = function() {
      var pack, _i, _len, _ref1;
      if (this.panels.size > 0) {
        return;
      }
      this.panelsByName = {};
      this.on('click', '.panels-menu li a, .panels-packages li a', (function(_this) {
        return function(e) {
          return _this.showPanel($(e.target).closest('li').attr('name'));
        };
      })(this));
      this.openDotAtom.on('click', function() {
        return atom.open({
          pathsToOpen: [atom.getConfigDirPath()]
        });
      });
      this.filterEditor.getEditor().on('contents-modified', (function(_this) {
        return function() {
          return _this.filterPackages();
        };
      })(this));
      this.addCorePanel('Settings', 'settings', function() {
        return new GeneralPanel;
      });
      this.addCorePanel('Keybindings', 'keyboard', function() {
        return new KeybindingsPanel;
      });
      this.addCorePanel('Packages', 'package', (function(_this) {
        return function() {
          return new PackagesPanel(_this.packageManager);
        };
      })(this));
      this.addCorePanel('Themes', 'paintcan', (function(_this) {
        return function() {
          return new ThemesPanel(_this.packageManager);
        };
      })(this));
      _ref1 = this.getPackages();
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        pack = _ref1[_i];
        this.addPackagePanel(pack);
      }
      if (this.panelToShow) {
        this.showPanel(this.panelToShow);
      }
      if (!this.activePanelName) {
        this.showPanel('Settings');
      }
      this.filterEditor.show();
      return this.filterEditor.redraw();
    };

    SettingsView.prototype.serialize = function() {
      var _ref1;
      return {
        deserializer: 'SettingsView',
        version: 2,
        activePanelName: (_ref1 = this.activePanelName) != null ? _ref1 : this.panelToShow,
        uri: this.uri
      };
    };

    SettingsView.prototype.getPackages = function() {
      var metadata, metadataPath, name, packageName, packagePath, _i, _len, _ref1, _ref2, _ref3;
      if (this.packages != null) {
        return this.packages;
      }
      this.packages = atom.packages.getLoadedPackages();
      _ref2 = (_ref1 = atom.config.get('core.disabledPackages')) != null ? _ref1 : [];
      for (_i = 0, _len = _ref2.length; _i < _len; _i++) {
        packageName = _ref2[_i];
        packagePath = atom.packages.resolvePackagePath(packageName);
        if (!packagePath) {
          continue;
        }
        if (metadataPath = CSON.resolve(path.join(packagePath, 'package'))) {
          try {
            metadata = CSON.readFileSync(metadataPath);
            name = (_ref3 = metadata != null ? metadata.name : void 0) != null ? _ref3 : packageName;
            if (!_.findWhere(this.packages, {
              name: name
            })) {
              this.packages.push({
                name: name,
                metadata: metadata,
                path: packagePath
              });
            }
          } catch (_error) {}
        }
      }
      this.packages.sort((function(_this) {
        return function(pack1, pack2) {
          var title1, title2;
          title1 = _this.packageManager.getPackageTitle(pack1);
          title2 = _this.packageManager.getPackageTitle(pack2);
          return title1.localeCompare(title2);
        };
      })(this));
      return this.packages;
    };

    SettingsView.prototype.redrawEditors = function() {
      var element, _i, _len, _ref1, _results;
      _ref1 = this.find('.editor');
      _results = [];
      for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
        element = _ref1[_i];
        _results.push($(element).view().redraw());
      }
      return _results;
    };

    SettingsView.prototype.addCorePanel = function(name, iconName, panel) {
      var panelMenuItem;
      panelMenuItem = $$(function() {
        return this.li({
          name: name
        }, (function(_this) {
          return function() {
            return _this.a({
              "class": "icon icon-" + iconName
            }, name);
          };
        })(this));
      });
      this.menuSeparator.before(panelMenuItem);
      return this.addPanel(name, panelMenuItem, panel);
    };

    SettingsView.prototype.addPackagePanel = function(pack) {
      var panelMenuItem;
      panelMenuItem = new PackageMenuView(pack, this.packageManager);
      this.panelPackages.append(panelMenuItem);
      return this.addPanel(pack.name, panelMenuItem, (function(_this) {
        return function() {
          return new InstalledPackageView(pack, _this.packageManager);
        };
      })(this));
    };

    SettingsView.prototype.addPanel = function(name, panelMenuItem, panelCreateCallback) {
      if (this.panelCreateCallbacks == null) {
        this.panelCreateCallbacks = {};
      }
      this.panelCreateCallbacks[name] = panelCreateCallback;
      if (this.panelToShow === name) {
        return this.showPanel(name);
      }
    };

    SettingsView.prototype.getOrCreatePanel = function(name) {
      var callback, panel, _ref1, _ref2;
      panel = (_ref1 = this.panelsByName) != null ? _ref1[name] : void 0;
      if (panel == null) {
        if (callback = (_ref2 = this.panelCreateCallbacks) != null ? _ref2[name] : void 0) {
          panel = callback();
          if (this.panelsByName == null) {
            this.panelsByName = {};
          }
          this.panelsByName[name] = panel;
          delete this.panelCreateCallbacks[name];
        }
      }
      return panel;
    };

    SettingsView.prototype.makePanelMenuActive = function(name) {
      this.sidebar.find('.active').removeClass('active');
      return this.sidebar.find("[name='" + name + "']").addClass('active');
    };

    SettingsView.prototype.showPanel = function(name) {
      var editorElement, index, panel, _i, _len, _ref1;
      if (panel = this.getOrCreatePanel(name)) {
        this.panels.children().hide();
        if (!$.contains(this.panels[0], panel[0])) {
          this.panels.append(panel);
        }
        panel.show();
        _ref1 = panel.find(".editor");
        for (index = _i = 0, _len = _ref1.length; _i < _len; index = ++_i) {
          editorElement = _ref1[index];
          $(editorElement).view().redraw();
        }
        panel.focus();
        this.makePanelMenuActive(name);
        this.activePanelName = name;
        return this.panelToShow = null;
      } else {
        return this.panelToShow = name;
      }
    };

    SettingsView.prototype.filterPackages = function() {
      var active, all, filterText;
      filterText = this.filterEditor.getEditor().getText();
      all = _.map(this.panelPackages.children(), function(item) {
        return {
          element: $(item),
          text: $(item).text()
        };
      });
      active = fuzzaldrin.filter(all, filterText, {
        key: 'text'
      });
      _.each(all, function(_arg) {
        var element;
        element = _arg.element;
        return element.hide();
      });
      return _.each(active, function(_arg) {
        var element;
        element = _arg.element;
        return element.show();
      });
    };

    SettingsView.prototype.removePanel = function(name) {
      var panel, _ref1;
      if (panel = (_ref1 = this.panelsByName) != null ? _ref1[name] : void 0) {
        panel.remove();
        delete this.panelsByName[name];
        return this.panelPackages.find("li[name=\"" + name + "\"]").remove();
      }
    };

    SettingsView.prototype.getTitle = function() {
      return "Settings";
    };

    SettingsView.prototype.getUri = function() {
      return this.uri;
    };

    SettingsView.prototype.isEqual = function(other) {
      return other instanceof SettingsView;
    };

    SettingsView.prototype.activatePackages = function(finishedCallback) {
      var iterator;
      iterator = function(pack, callback) {
        var error;
        try {
          return pack.activateConfig();
        } catch (_error) {
          error = _error;
          return console.error("Error activating package config for \u201C" + pack.name + "\u201D", error);
        } finally {
          callback();
        }
      };
      return async.each(atom.packages.getLoadedPackages(), iterator, finishedCallback);
    };

    return SettingsView;

  })(ScrollView);

}).call(this);
