// package-generator/lib/package-generator-view.js
(function() {
  var $, BufferedProcess, EditorView, PackageGeneratorView, View, fs, path, _, _ref,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  path = require('path');

  _ = require('underscore-plus');

  _ref = require('atom'), $ = _ref.$, BufferedProcess = _ref.BufferedProcess, EditorView = _ref.EditorView, View = _ref.View;

  fs = require('fs-plus');

  module.exports = PackageGeneratorView = (function(_super) {
    __extends(PackageGeneratorView, _super);

    function PackageGeneratorView() {
      return PackageGeneratorView.__super__.constructor.apply(this, arguments);
    }

    PackageGeneratorView.prototype.previouslyFocusedElement = null;

    PackageGeneratorView.prototype.mode = null;

    PackageGeneratorView.content = function() {
      return this.div({
        "class": 'package-generator overlay from-top'
      }, (function(_this) {
        return function() {
          _this.subview('miniEditor', new EditorView({
            mini: true
          }));
          _this.div({
            "class": 'error',
            outlet: 'error'
          });
          return _this.div({
            "class": 'message',
            outlet: 'message'
          });
        };
      })(this));
    };

    PackageGeneratorView.prototype.initialize = function() {
      atom.workspaceView.command("package-generator:generate-package", (function(_this) {
        return function() {
          return _this.attach('package');
        };
      })(this));
      atom.workspaceView.command("package-generator:generate-syntax-theme", (function(_this) {
        return function() {
          return _this.attach('theme');
        };
      })(this));
      this.miniEditor.hiddenInput.on('focusout', (function(_this) {
        return function() {
          return _this.detach();
        };
      })(this));
      this.on('core:confirm', (function(_this) {
        return function() {
          return _this.confirm();
        };
      })(this));
      return this.on('core:cancel', (function(_this) {
        return function() {
          return _this.detach();
        };
      })(this));
    };

    PackageGeneratorView.prototype.attach = function(mode) {
      this.mode = mode;
      this.previouslyFocusedElement = $(':focus');
      this.message.text("Enter " + mode + " path");
      atom.workspaceView.append(this);
      if (this.mode === 'package') {
        this.setPathText("my-package");
      } else {
        this.setPathText("my-theme-syntax", [0, 8]);
      }
      return this.miniEditor.focus();
    };

    PackageGeneratorView.prototype.setPathText = function(placeholderName, rangeToSelect) {
      var editor, endOfDirectoryIndex, packagesDirectory, pathLength;
      editor = this.miniEditor.editor;
      if (rangeToSelect == null) {
        rangeToSelect = [0, placeholderName.length];
      }
      packagesDirectory = this.getPackagesDirectory();
      editor.setText(path.join(packagesDirectory, placeholderName));
      pathLength = editor.getText().length;
      endOfDirectoryIndex = pathLength - placeholderName.length;
      return editor.setSelectedBufferRange([[0, endOfDirectoryIndex + rangeToSelect[0]], [0, endOfDirectoryIndex + rangeToSelect[1]]]);
    };

    PackageGeneratorView.prototype.detach = function() {
      var _ref1;
      if (!this.hasParent()) {
        return;
      }
      if ((_ref1 = this.previouslyFocusedElement) != null) {
        _ref1.focus();
      }
      return PackageGeneratorView.__super__.detach.apply(this, arguments);
    };

    PackageGeneratorView.prototype.confirm = function() {
      if (this.validPackagePath()) {
        return this.createPackageFiles((function(_this) {
          return function() {
            var packagePath;
            packagePath = _this.getPackagePath();
            atom.open({
              pathsToOpen: [packagePath]
            });
            return _this.detach();
          };
        })(this));
      }
    };

    PackageGeneratorView.prototype.getPackagePath = function() {
      var packageName, packagePath;
      packagePath = this.miniEditor.getText();
      packageName = _.dasherize(path.basename(packagePath));
      return path.join(path.dirname(packagePath), packageName);
    };

    PackageGeneratorView.prototype.getPackagesDirectory = function() {
      return _.last(atom.packages.getPackageDirPaths());
    };

    PackageGeneratorView.prototype.validPackagePath = function() {
      if (fs.existsSync(this.getPackagePath())) {
        this.error.text("Path already exists at '" + (this.getPackagePath()) + "'");
        this.error.show();
        return false;
      } else {
        return true;
      }
    };

    PackageGeneratorView.prototype.initPackage = function(packagePath, callback) {
      return this.runCommand(atom.packages.getApmPath(), ['init', "--" + this.mode, "" + packagePath], callback);
    };

    PackageGeneratorView.prototype.linkPackage = function(packagePath, callback) {
      return this.runCommand(atom.packages.getApmPath(), ['link', "" + packagePath], callback);
    };

    PackageGeneratorView.prototype.createPackageFiles = function(callback) {
      var packagePath, packagesDirectory;
      packagePath = this.getPackagePath();
      packagesDirectory = this.getPackagesDirectory();
      if (packagePath.indexOf(path.join(packagesDirectory, path.sep)) === 0) {
        return this.initPackage(packagePath, callback);
      } else {
        return this.initPackage(packagePath, (function(_this) {
          return function() {
            return _this.linkPackage(packagePath, callback);
          };
        })(this));
      }
    };

    PackageGeneratorView.prototype.runCommand = function(command, args, exit) {
      return new BufferedProcess({
        command: command,
        args: args,
        exit: exit
      });
    };

    return PackageGeneratorView;

  })(View);

}).call(this);
